### PR TITLE
[Flang][Clang] Add support for frame pointers in Flang Driver

### PR DESCRIPTION
**Pull request by Radu2k**
Add support for frame pointers in Flang.

### CODE REVIEW

**Review by Radu2k — COMMENTED** [review]
Small fix.

**Comment by banach-space** [comment]
Could you add some tests? In e.g. "driver-help.f90" we merely test that a flag is visible.

File: clang/lib/Driver/ToolChains/CommonArgs.cpp [hunk 1]
```diff
@@ -960,7 +1099,7 @@ bool tools::addOpenMPRuntime(ArgStringList &CmdArgs, const ToolChain &TC,
     CmdArgs.push_back("-Bdynamic");
 
   if (RTKind == Driver::OMPRT_GOMP && GompNeedsRT)
-      CmdArgs.push_back("-lrt");
+    CmdArgs.push_back("-lrt");
```
Comment (banach-space) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:1102]: This and the following changes in this file are unrelated and merely formatting fixes. Please, could you reduce the noise and revert these? I am happy for you to submit them as a separate PR.

**Comment by banach-space** [comment]
> Could you add some tests? In e.g. "driver-help.f90" we merely test that a flag is visible.

I see that you added [frame-pointer-codegen.f90](https://github.com/llvm/llvm-project/pull/72146/files#diff-76ca50365c8a0eb4797ce10b121021e7da52cc86523d806e1ade530a897de7e8), but that does not test how flags added in this patch impact code-gen. Please add a test that will demonstrate how this flag impacts the generated code.

File: clang/include/clang/Driver/Options.td [hunk 2]
```diff
@@ -6752,10 +6754,7 @@ def new_struct_path_tbaa : Flag<["-"], "new-struct-path-tbaa">,
 def mdebug_pass : Separate<["-"], "mdebug-pass">,
   HelpText<"Enable additional debug output">,
   MarshallingInfoString<CodeGenOpts<"DebugPass">>;
-def mframe_pointer_EQ : Joined<["-"], "mframe-pointer=">,
-  HelpText<"Specify which frame pointers to retain.">, Values<"all,non-leaf,none">,
-  NormalizedValuesScope<"CodeGenOptions::FramePointerKind">, NormalizedValues<["All", "NonLeaf", "None"]>,
-  MarshallingInfoEnum<CodeGenOpts<"FramePointer">, "None">;
+
```
Comment (tblah) [inline clang/include/clang/Driver/Options.td:6757]: nit: did you mean to add this extra line?

File: flang/lib/Frontend/CompilerInvocation.cpp [hunk 3]
```diff
@@ -1114,7 +1114,8 @@ bool CompilerInvocation::createFromArgs(
     res.loweringOpts.setLowerToHighLevelFIR(false);
   }
 
-  if (args.hasArg(clang::driver::options::OPT_flang_experimental_polymorphism)) {
+  if (args.hasArg(
```
Comment (tblah) [inline flang/lib/Frontend/CompilerInvocation.cpp:1117]: This looks unrelated to your patch

**Review by tblah — CHANGES_REQUESTED** [review]
While I guess it isn't wrong, it feels weird do declare all of the storage on the front end driver side, but never set them (the member of the CodeGenOpttions class, the changes in Frontend/CodeGenOptions.h, adding it to MLIRToLLVMPassPipelineConfig).

This might be a bit nitpicky but I would either remove these changes or add frontend driver code to read the argument and set it in the MLIRToLLVMPassPipelineConfig and CodeGenOptions (although this would be untestable so others may disagree).

With regard to the front end driver stuff, I don't know if we gain anything by adding a flang specific definition of the enum in CodeGenOptions.h, as we will just convert to an llvm::FramePointerKind to populate MLIRToLLVMPassPipelineConfig. If you're sure that the llvm version is the one you want to use in MLIRToLLVMPassPipelineConfig, then just use that same type in flang::CodeGenOptions too.

File: flang/include/flang/Frontend/CodeGenOptions.h [hunk 4]
```diff
@@ -49,6 +49,26 @@ class CodeGenOptionsBase {
...
+      return "none";
+    case FramePointerKind::NonLeaf:
+      return "non-leaf";
+    case FramePointerKind::All:
+      return "all";
+    }
+
+    llvm_unreachable("invalid FramePointerKind");
+  };
+
```
Comment (kiranchandramohan) [inline flang/include/flang/Frontend/CodeGenOptions.h:71]: Move this code and the same code in `clang/include/clang/Basic/CodeGenOptions.h` to llvm/include/llvm/Frontend/Driver.

Comment (Radu2k) [reply flang/include/flang/Frontend/CodeGenOptions.h:71]: Issue has been addressed

File: flang/include/flang/Frontend/CodeGenOptions.h [hunk 5]
```diff
@@ -49,6 +49,7 @@ class CodeGenOptionsBase {
 class CodeGenOptions : public CodeGenOptionsBase {
 
 public:
+
```
Comment (tblah) [inline flang/include/flang/Frontend/CodeGenOptions.h:52]: nit: unrelated change

**Review by tblah — COMMENTED** [review]
Thanks for the update. See my previous comment about the changes to `MLIRToLLVMPassPipelineConfig` and `CodeGenOptions.def`.

**Review by tblah — APPROVED** [review]
LGTM, thanks!

**Comment by tblah** [comment]
Merged manually https://github.com/llvm/llvm-project/commit/ea4eb691f4955e3b784ebf9bc94a47186838c6f2
